x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2867 @@
+
+```yaml
+#  [... file continues above...]
+
+ResponseFields:
+  type: object
+  properties:
+    id:
+      $ref: './common_attributes.schema.yaml#/components/schemas/RuleObjectId'
+    rule_id:
+      $ref: './common_attributes.schema.yaml#/components/schemas/RuleSignatureId'
+    immutable:
+      $ref: './common_attributes.schema.yaml#/components/schemas/IsRuleImmutable'
+    rule_source:
+      $ref: './common_attributes.schema.yaml#/components/schemas/RuleSource'
+    #  [...  more response fields ...]
+  required:
+    - id
+    - rule_id
+    - immutable
+    - updated_at
+    - updated_by
+    - created_at
+    - created_by
+    - revision
+    - rule_source
+#  [... file continues below...]
+```
+
+##### Rule Import request schema
+
+We also need to modify the `RuleToImport` schema, since now we will be allowing the importing of both custom rules and prebuilt rules.
+
+Currently, `RuleToImport` optionally accepts the `immutable` param, but rejects validation if its value is set to anything else than `false` - since we don't currently support importing prebuilt rules. 
+
+We will be changing the mechanism for importing rules so that, besides the `rule_id`, the `version` of the rule is also a required parameter. These two parameters will be used to determine if the rule is prebuilt or not, and dynamically calculate `rule_source` during import.
+
+See the detailed explanation for this mechanism in the [Exporting and importing rules](#exporting-and-importing-rules) sections.
+
+The `immutable` and `rule_source` fields will be ignored if passed in the request payload.
+
+_Source: [x-pack/plugins/security_solution/common/api/detection_engine/rule_management/import_rules/rule_to_import.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/api/detection_engine/rule_management/import_rules/rule_to_import.ts)_
+
+```ts
+// [... file continues above...]
+import {
+  // [...]
+  RuleSignatureId,
+  IsRuleImmutable,
+  ExternalSourceAttributes,
+  RuleVersion,
+} from '../../model/rule_schema';
+
+export type RuleToImport = z.infer<typeof RuleToImport>;
+export type RuleToImportInput = z.input<typeof RuleToImport>;
+export const RuleToImport = BaseCreateProps.and(TypeSpecificCreateProps).and(
+  ResponseFields.partial().extend({
+    rule_id: RuleSignatureId,
+    version: RuleVersion,
+  })
+);
+```
+
+#### Internal rule schema
+
+**The internal rule schema** needs to represent that the new `ruleSource` field may not always exist, so both need to be optional.
+
+_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts)_
+
+```ts
+export type BaseRuleParams = z.infer<typeof BaseRuleParams>;
+export const BaseRuleParams = z.object({
+  // [...]
+
+  immutable: IsRuleImmutable,
```

Comment:
Should this become optional in the internal schema? Will newly created rules have this param stored in ES?

Reply:
@jpdjere I'm not strongly against this, but I'm not sure I understand why we'd need to store `immutable` in those rules that have been migrated-on-write or newly created. We will have normalization-on-read + fallback to `immutable` in the `find` endpoint + the field will be deprecated once we release prebuilt rule customization. If you suggest to store it anyway, then when do you think we will stop storing it?

Reply:
My original idea was: "remove it only when we release a new API version".

But now I think that what you propose is actually tidier and takes care of the problem now instead of later (like in two years, or so.)

Reply:
- Reflect this decision on the Migration-on-write section
- Update the table that shows migration scenarios